### PR TITLE
[CONTINT-3539] Add support of container metrics to kata containers

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR adds support of container metrics to kata containers using the generic metrics provider.

Kata containers can't be monitored by the system collector as cgroups are not visible from the host. It means we should fall back to the containerd/cri collector for container metrics.

### Motivation
We would like to monitor kata containers.

### Additional Notes
Note that containerd can't retrieve IO and network metrics. We need to fallback on the Kubelet collector for this. Another PR will fix this.

### Possible Drawbacks / Trade-offs
N/A

### Describe how to test/QA your changes
Setup a kata container environment. You'll need:
1. A VM with nested virtualization or a baremetal linux machine
See https://github.com/kata-containers/kata-containers/blob/main/docs/install/gce-installation-guide.md and only create the VM. Then follow the steps here to enable KVM https://phoenixnap.com/kb/ubuntu-install-kvm. Restart your VM from the google cloud console.

3. Install minikube and a dummy app with kata. For this, follow https://github.com/kata-containers/kata-containers/blob/main/docs/install/minikube-installation-guide.md

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by drichards-87 — APPROVED** [review]

File: pkg/util/containers/metrics/containerd/collector.go [hunk 1]
```diff
@@ -75,10 +76,17 @@ func newContainerdCollector(cache *provider.Cache) (provider.CollectorMetadata,
 		ContainerIDForPID: provider.MakeRef[provider.ContainerIDForPIDRetriever](collector, collectorPriority),
 	}
 
+	kataCollectors := &provider.Collectors{
+		Stats:             provider.MakeRef[provider.ContainerStatsGetter](collector, collectorPriority),
+		PIDs:              provider.MakeRef[provider.ContainerPIDsGetter](collector, collectorPriority),
+		ContainerIDForPID: provider.MakeRef[provider.ContainerIDForPIDRetriever](collector, collectorPriority),
```
Comment (wdhif) [inline pkg/util/containers/metrics/containerd/collector.go:82]: This would not work with Kata containers right?

Comment (AliDatadog) [reply pkg/util/containers/metrics/containerd/collector.go:82]: When we inspect a pod, a pid is reported. I think it corresponds to the pid of the virtual machine. So technically it still gives a mapping from pid to ContainerID.

But it is true that we most likely won't have to use it given that dogstatsd is the only caller and it requires to mount the UDS socket, which is not supported by kata containers anyway

File: comp/core/workloadmeta/collectors/util/container.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+package util
+
+import (
+	"github.com/DataDog/datadog-agent/comp/core/workloadmeta"
+)
+
+// ExtractRuntimeFlavor extracts the runtime from a runtime string.
+func ExtractRuntimeFlavor(runtime string) workloadmeta.ContainerRuntimeFlavor {
+	if runtime == "io.containerd.kata.v2" {
+		return workloadmeta.ContainerRuntimeFlavorKata
```
Comment (vboulineau) [inline comp/core/workloadmeta/collectors/util/container.go:16]: It's only related to `containerd` and, as such, it should go to `containerd` collector.

File: pkg/util/containers/metrics/containerd/collector.go (hunk 1 above)
Comment (vboulineau) [reply pkg/util/containers/metrics/containerd/collector.go:82]: I think you can only register `Stats`

File: comp/core/workloadmeta/collectors/util/container.go (hunk 2 above)
Comment (AliDatadog) [reply comp/core/workloadmeta/collectors/util/container.go:16]: My bad I did a rm and not a `git rm`.

File: pkg/util/containers/metrics/kubelet/collector.go [hunk 3]
```diff
@@ -78,9 +78,10 @@ func newKubeletCollector(*provider.Cache) (provider.CollectorMetadata, error) {
 	return provider.CollectorMetadata{
 		ID: collectorID,
 		Collectors: provider.CollectorCatalog{
-			provider.RuntimeNameContainerd: collectors,
-			provider.RuntimeNameCRIO:       collectors,
-			provider.RuntimeNameDocker:     collectors,
+			provider.NewRuntimeMetadata(string(provider.RuntimeNameContainerd), ""):                                 collectors,
+			provider.NewRuntimeMetadata(string(provider.RuntimeNameContainerd), string(provider.RuntimeFlavorKata)): collectors,
```
Comment (AliDatadog) [inline pkg/util/containers/metrics/kubelet/collector.go:82]: I registered stats and network for the kubelet

File: pkg/util/containers/metrics/provider/provider.go [hunk 4]
```diff
@@ -65,9 +65,41 @@ var (
...
+// NewRuntimeMetadata returns a new RuntimeMetadata
+func NewRuntimeMetadata(runtime, flavor string) RuntimeMetadata {
+	return RuntimeMetadata{
+		flavor:  RuntimeFlavor(flavor),
+		runtime: Runtime(runtime),
+	}
+}
+
+// String returns the runtime compose.
+func (r *RuntimeMetadata) String() Runtime {
```
Comment (vboulineau) [inline pkg/util/containers/metrics/provider/provider.go:92]:
```suggestion
func (r *RuntimeMetadata) String() string {
```

File: pkg/util/containers/metrics/provider/provider.go [hunk 5]
```diff
@@ -65,9 +65,41 @@ var (
...
+	if r.flavor == RuntimeFlavorKata {
+		return r.runtime + Runtime("-") + Runtime(RuntimeFlavorKata)
```
Comment (vboulineau) [inline pkg/util/containers/metrics/provider/provider.go:94]: I don't think you need to be specific. If `r.flavor != nil` you concatenate, if not you return `r.runtime`.

File: pkg/util/containers/metrics/provider/provider.go [hunk 6]
```diff
@@ -65,9 +65,41 @@ var (
...
+func NewRuntimeMetadata(runtime, flavor string) RuntimeMetadata {
+	return RuntimeMetadata{
+		flavor:  RuntimeFlavor(flavor),
+		runtime: Runtime(runtime),
+	}
+}
+
+// String returns the runtime compose.
+func (r *RuntimeMetadata) String() string {
+	if r.flavor == "" {
```
Comment (vboulineau) [inline pkg/util/containers/metrics/provider/provider.go:93]:
```suggestion
	if r.flavor != "" {
```

File: pkg/util/containers/metrics/provider/provider.go [hunk 7]
```diff
@@ -65,9 +65,41 @@ var (
...
+		return string(r.runtime) + "-" + string(r.runtime)
```
Comment (vboulineau) [inline pkg/util/containers/metrics/provider/provider.go:94]:
```suggestion
		return string(r.runtime) + "-" + string(r.flavor)
```

**Review by wdhif — APPROVED** [review]

**Review by kkhor-datadog — APPROVED** [review]

**Pull request merged**